### PR TITLE
Implement open()

### PR DESCRIPTION
**Pull request by brettcannon**

### CODE REVIEW

File: importlib_resources/__init__.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+from typing.io import BinaryIO
+
+if sys.version_info >= (3, 6):
+    Path = Union[str, os.PathLike]
+else:
+    Path = str
+
+
+def _get_package(package_name):
+    module = importlib.import_module(package_name)
```
Comment (warsaw) [inline importlib_resources/__init__.py]: I realize we haven't discussed this, but I wonder if we should accept either a package name or a module/package object?  I suppose it's also not that difficult to just pass in the `__name__`.

File: importlib_resources/__init__.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    if module.__spec__.submodule_search_locations is None:
```
Comment (warsaw) [inline importlib_resources/__init__.py]: Are the attributes of `__spec__` defined anywhere but the code?  There's [this mention](https://docs.python.org/3/reference/import.html?highlight=__spec__#__spec__) but it provides no details.  [PEP 451](https://www.python.org/dev/peps/pep-0451/) has more details, but PEPs aren't documentation.

File: importlib_resources/__init__.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    if os.path.isabs(path):
+        raise ValueError("{!r} is absolute".format(path))
+    normalized_path = os.path.normpath(path)
+    if normalized_path.startswith(".."):
+        raise ValueError("{!r} attempts to traverse past package".format(path))
+    else:
+        return normalized_path
+
+
+def open(package_name: str, path: Path) -> BinaryIO:
```
Comment (warsaw) [inline importlib_resources/__init__.py]: I know we talked about this (though we didn't capture it in an issue), but now I'm waffling. ;)

I'm not so sure about accepting additional `path` elements, or whether we should limit this to an unresolved name.  This is related to whether you should be required to traverse into the resource's location solely by the package path.  I almost always make my `data` directories packages, so I would tend to use

`with open('mailman.rest.tests.data', 'request.json'):` or some such.  But that does require me to put an `__init__py` in `mailman/rest/tests/data`.  That's fine with me though.

The implication here though is that `data` wouldn't have to be a subpackage and I could write:

`with open('mailman.rest.tests', 'data/requests.json'):`

Let's discuss (in an issue if necessary)!

File: importlib_resources/__init__.py (hunk 2 above)
Comment (brettcannon) [reply importlib_resources/__init__.py]: There's https://docs.python.org/3/library/importlib.html#importlib.machinery.ModuleSpec

File: importlib_resources/__init__.py (hunk 1 above)
Comment (brettcannon) [reply importlib_resources/__init__.py]: We actually did discuss this and decided it was easier to just take a string. If you want to open the discussion again we can. I assume we would do `hasattr(package, '__spec__')` to get the package's name.

Comment (brettcannon) [reply importlib_resources/__init__.py]: https://github.com/brettcannon/importlib_resources/issues/7

File: importlib_resources/__init__.py (hunk 3 above)
Comment (brettcannon) [reply importlib_resources/__init__.py]: https://github.com/brettcannon/importlib_resources/issues/8

File: importlib_resources/__init__.py (hunk 2 above)
Comment (brettcannon) [reply importlib_resources/__init__.py]: But yes, there isn't a listing of the attribute in the language spec itself (probably wouldn't hurt if you want to open an issue).

Comment (warsaw) [reply importlib_resources/__init__.py]: I'm [way ahead of you](https://github.com/python/cpython/pull/4010) :)

**Comment by brettcannon** [comment]
@warsaw this should be ready for review again.

**Pull request merged**